README.md
# MLR-Legacy-Transformer
[![Build Status](https://travis-ci.org/USGS-CIDA/MLR-Legacy-Transformer.svg?branch=master)](https://travis-ci.org/USGS-CIDA/MLR-Legacy-Transformer)
[![Coverage Status](https://coveralls.io/repos/github/USGS-CIDA/MLR-Legacy-Transformer/badge.svg)](https://coveralls.io/github/USGS-CIDA/MLR-Legacy-Transformer)

## Service Description
This service is part of the MLR microservices and is responsible for performing various data transformations that must happen to an incoming location before that location is persisted in the database. Due to the way that DDot files are produced there are several pieces of data transmitted in the DDot file that are not necessarily in the proper format to be directly persisted into the Database. There are various software tools that are responsible for creating DDot files and all of them are older and not regularly maintained so making changes to each of them to perform these transformations was not an option. As a result, this service was created to be a catch-all for any transformations that we would have to make to incoming data (not all transformations were known when this srevice was created). As things went on it turned out that there were only two minor transformations that had to be done to DDot files, and as such this service is now being considered for removal and having the transformation logic performed in the Gateway or another service.

Currently the service performs the following two transformations:

1. Converting the raw station name into a normalized station name. This transformation removes certain special characters from the raw station name and normalizes the format and casing. This allows us to limit the number of inexact-duplicate station names in the database and ensure that we keep station names unique.

2. Converting the format of the Latitude and Longitude station coordinates to a decimal format. The database, and downstream applications which read from it, expects to store these coordinates in a decimal format but the DDot creation tools output these in a degree-minute-second format.

These transformations must occur before the duplicate validation performed by the Legacy CRU service as the normalized station name is not sent in with DDot file and it is used for that validation process.

The API methods for these transformations including expected inputs and result outputs can be examined in more detail in the Swagger API documentation for this service.

## Building and Running

This project has been built and tested with python 3.6.x. To build the project locally you will need
python 3 and virtualenv installed.
```bash
% virtualenv --python=python3 env
% env/bin/pip install -r requirements.txt
```
To run the tests:
```bash
env/bin/python -m unittest
```

To run the application locally execute the following:
```bash
% env/bin/python app.py
```

The swagger documentation can then be accessed at http://127.0.0.1:5000/api (when running locally using the above command).

Default configuration variables can be overridden be creating a .env file. For instance to turn debug on, 
you will want to create an .env with the following:
```python
DEBUG = True
```

For local development, you will need to provide a JWT token to the service. This can be done through the Swagger 
documents by clicking the Authorize button and entering 'Bearer your.jwt.token'.

You can use a valid JWT token generated by another service. You will need to set it's JWT_PUBLIC_KEY to the public 
key used to generate the token, as well as the JWT_DECODE_AUDIENCE (if any) and the JWT_ALGORITHM 
(if different than RS256). If you don't want to verify the cert on this service, set AUTH_CERT_PATH to False.

Alternatively, you can generate your own token by using the python package jwt. In the python interpreter, do the following
```python
import jwt
jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret', algorithm='HS256')
```

The output of this command will be the token that you can use. You will need to set JWT_SECRET_KEY to 'secret' in 
your local .env file. See http://flask-jwt-simple.readthedocs.io/en/latest/options.html for the other options that 
you can use.

## Running with Docker 
This application can also be run locally using the docker container built during the build process, though this does not allow the application to be run in debug mode. The included `docker-compose` file has 2 profiles to choose from when running the application locally:

1. mlr-legacy-transformer: This is the default profile which runs the application as it would be in our cloud environment. This is not recommended for local development as it makes configuring connections to other services running locally on your machine more difficult.
2. mlr-legacy-transformer-local-dev: This is the profile which runs the application as it would be in the mlr-local-dev project, and is configured to make it easy to replace the mlr-legacy-transformer instance in the local-dev project with this instance. It is run the same as the `mlr-legacy-transformer` profile, except it uses the docker host network driver.

Before any of these options are able to be run you must also generate certificates for this application to serve using the `create_certificates` script in the `docker/certificates` directory. Additionally, this service must be able to connect to a running instance of Water Auth when starting, and it is recommended that you use the Water Auth instance from the `mlr-local-dev` project to accomplish this. In order for this application to communicate with any downstream services that it must call, including Water Auth, you must also place the certificates that are being served by those services into the `docker/certificates/import_certs` directory to be imported into the Python CA Certificates of the running container.

To build and run the application after completing the above steps you can run: `docker-compose up --build {profile}`, replacing `{profile}` with one of the options listed above.

The swagger documentation can then be accessed at http://127.0.0.1:6020/api

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.10.0] - 2019-04-12 - End of Pilot
### Added
- Docker configuration

### Changed
- Merged this repository with the mlr-legacy-transformer-docker repository

## [0.9.0] - 2019-03-01
## Changed
- Standardized error response format.

## [0.8.0] - 2019-01-31
## Changed
- update `Flask` to 0.12.3
- update `urllib3`  to 1.23

## [0.7.0] - 2018-11-27
## Added
- warn logging for lat/lon transformation errors.

## Changed
- Update `requests` to 2.20.0

## [0.6.0] - 2018-08-23
### Updated
- isuftin@usgs.gov - Updated the version constraint for pyca/cryptography due to
CVE https://nvd.nist.gov/vuln/detail/CVE-2018-10903

### Added
- Dockerfile Healthcheck

### Removed
- Dockerfile
- Dockerfile-DOI
- gunicorn_config.py

## [0.5.0] - 2017-11-20
### Added
- GET endpoint /version to show the current version and artifact name
- Authentication for /transformer/decimal_location and transformer/station_ix endpoints
- HTTPS Support

## [0.4.0] - 2017-11-01
### Change
- Trim whitespace from coordinateDatumCode before using it as a key to determine what projection to use.

## [0.3.0] - 2017-10-18
### Change
- Handle bad latitude/longitude values including out of range values by return null values for decimalLatitude and decimalLongitude.

## 0.2.0 - 2017-10-04
### Add
- POST endpoint /transformer/decimal_location takes a json object containing string latitude, longitude, and coordinateDatumCode and returns
decimalLatitude and decimalLongitude. Values will be null if invalid latitude, longitude, or coordinateDatumCode
- POST endpoint /transformer/station_ix takes a json object containing string station name and responds with stationIx.
- Swagger documention endpoint /api

[Unreleased]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.10.0...master
[0.10.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.9.0...MLR-Legacy-Transformer-0.10.0
[0.9.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.8.0...MLR-Legacy-Transformer-0.9.0
[0.8.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.7.0...MLR-Legacy-Transformer-0.8.0
[0.7.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.6.0...MLR-Legacy-Transformer-0.7.0
[0.6.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.5.0...MLR-Legacy-Transformer-0.6.0
[0.5.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.4.0...MLR-Legacy-Transformer-0.5.0
[0.4.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.3.0...MLR-Legacy-Transformer-0.4.0
[0.3.0]: https://github.com/USGS-CIDA/MLR-Legacy-Transformer/compare/MLR-Legacy-Transformer-0.2.0...MLR-Legacy-Transformer-0.3.0


tests/test_utils.py

from unittest import TestCase

from utils import transform_to_decimal_degrees, transform_latitude_to_decimal_degrees, \
    transform_longitude_to_decimal_degrees, transform_location_to_decimal_location

class TestTransformToDecimalDegrees(TestCase):

    def test_zero_values(self):
        self.assertEqual(transform_to_decimal_degrees('00', '00', '00.0'), 0)
        self.assertEqual(transform_to_decimal_degrees('-000', '00', '00'), 0)

    def test_without_decimal_seconds(self):
        self.assertAlmostEqual(transform_to_decimal_degrees(' 97', '43', '12'), 97.72, 2)
        self.assertAlmostEqual(transform_to_decimal_degrees('-123', '43', '49'), -123.73027778, 8)

    def test_with_decimal_seconds(self):
        self.assertAlmostEqual(transform_to_decimal_degrees(' 97', '43', '12.56'), 97.72015556, 8)
        self.assertAlmostEqual(transform_to_decimal_degrees('-123', '43', '49.231'), -123.730341, 5)


class TestTransformLatitudeToDecimalDegrees(TestCase):

    def test_zero_values(self):
        self.assertEqual(transform_latitude_to_decimal_degrees(' 000000'), 0)

    def test_with_positive_latitude(self):
        self.assertAlmostEqual(transform_latitude_to_decimal_degrees(' 474312'), 47.72, 2)
        self.assertAlmostEqual(transform_latitude_to_decimal_degrees(' 085623'), 8.93972222, 8)

    def test_with_negative_latitude(self):
        self.assertAlmostEqual(transform_latitude_to_decimal_degrees('-474312'), -47.72, 2)
        self.assertAlmostEqual(transform_latitude_to_decimal_degrees('-085623'), -8.93972222, 8)

    def test_with_outof_range_latitude(self):
        with self.assertRaises(ValueError):
            transform_latitude_to_decimal_degrees('400000')


class TestTransformLongitudeToDecimalDegrees(TestCase):

    def test_zero_values(self):
        self.assertEqual(transform_longitude_to_decimal_degrees(' 000000'), 0)

    def test_with_positive_latitude(self):
        self.assertAlmostEqual(transform_longitude_to_decimal_degrees(' 0974312'), -97.72, 2)
        self.assertAlmostEqual(transform_longitude_to_decimal_degrees(' 1234349'), -123.73027778, 8)

    def test_with_negative_latitude(self):
        self.assertAlmostEqual(transform_longitude_to_decimal_degrees('-0974312'), 97.72, 2)
        self.assertAlmostEqual(transform_longitude_to_decimal_degrees('-1234349'), 123.73027778, 8)

    def test_with_out_of_range_longitude(self):
        with self.assertRaises(ValueError):
            transform_longitude_to_decimal_degrees('1000000')


class TestTransformLocationTo_DecimalLocation(TestCase):

    def test_with_unconvertable_datum(self):
        self.assertEqual(transform_location_to_decimal_location(' 374312    ', '-1234349    ', 'WGS84     '),
                         {'decimalLatitude' : None, 'decimalLongitude': None})

    def test_with_nad83_datum(self):
        result = transform_location_to_decimal_location(' 374312    ', ' 1234349    ', 'NAD83     ')
        self.assertAlmostEqual(result.get('decimalLatitude'), 37.72, 5)
        self.assertAlmostEqual(result.get('decimalLongitude'), -123.73027778, 8)

        result = transform_location_to_decimal_location(' 374312    ', ' 1234349    ', 'NAD83')
        self.assertAlmostEqual(result.get('decimalLatitude'), 37.72, 5)
        self.assertAlmostEqual(result.get('decimalLongitude'), -123.73027778, 8)

    def test_with_nad27_datum(self):
        result = transform_location_to_decimal_location(' 383009    ', ' 0814256    ', 'NAD27     ')
        self.assertAlmostEqual(result.get('decimalLatitude'), 38.5025922044846, 8)
        self.assertAlmostEqual(result.get('decimalLongitude'), -81.7154066249968, 8)

        result = transform_location_to_decimal_location(' 383009    ', ' 0814256    ', 'NAD27')
        self.assertAlmostEqual(result.get('decimalLatitude'), 38.5025922044846, 8)
        self.assertAlmostEqual(result.get('decimalLongitude'), -81.7154066249968, 8)

    def test_with_non_numeric_lat_lon(self):
        self.assertEqual(transform_location_to_decimal_location('AAA',  ' 0814256    ', 'NAD27     '),
                         {'decimalLatitude': None, 'decimalLongitude': None})
        self.assertEqual(transform_location_to_decimal_location(' 374312    ', 'AAA', 'NAD27     '),
                         {'decimalLatitude': None, 'decimalLongitude': None})

    def test_with_invalid_numeric_lat_lon(self):
        self.assertEqual(transform_location_to_decimal_location('400000', ' 0814256    ', 'NAD83     '),
                         {'decimalLatitude': None, 'decimalLongitude': None})
        self.assertEqual(transform_location_to_decimal_location(' 400000', '1000000', 'NAD83     '),
                         {'decimalLatitude': None, 'decimalLongitude': None})





tests/test_services.py

import json
from unittest import TestCase, mock

import jwt

import app


@mock.patch('services.transform_location_to_decimal_location')
class DecimalLocationTestCase(TestCase):

    def setUp(self):
        app.application.config['JWT_SECRET_KEY'] = 'secret'
        app.application.config['JWT_PUBLIC_KEY'] = None
        app.application.config['JWT_ALGORITHM'] = 'HS256'
        app.application.config['AUTH_TOKEN_KEY_URL'] = ''
        app.application.config['JWT_DECODE_AUDIENCE'] = None
        app.application.testing = True
        self.app_client = app.application.test_client()

    def test_good_request(self, mock):
        good_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret')
        mock.return_value = {'decimalLatitude' : 40, 'decimalLongitude': -100}
        response = self.app_client.post('/transformer/decimal_location',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(good_token.decode('utf-8'))},
                                        data=json.dumps({
                                            'latitude': ' 400000    ',
                                            'longitude' : ' 1000000    ',
                                            'coordinateDatumCode' : 'NAD27      '}))
        self.assertEqual(response.status_code, 200)
        self.assertEqual(json.loads(response.data), {'decimalLatitude' : 40, 'decimalLongitude': -100})

    def test_missing_keys(self, mock):
        good_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret')
        response = self.app_client.post('/transformer/decimal_location',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(good_token.decode('utf-8'))},
                                        data=json.dumps({
                                            'latitudee': ' 400000    ',
                                            'longitude' : ' 1000000    ',
                                            'coordinateDatumCode' : 'NAD27      '}))
        self.assertEqual(response.status_code, 400)
        self.assertIn(b'latitude', response.data)

    def test_no_auth_header(self, mock):
        response = self.app_client.post('/transformer/decimal_location',
                                        content_type='application/json',
                                        data=json.dumps({
                                            'latitudee': ' 400000    ',
                                            'longitude': ' 1000000    ',
                                            'coordinateDatumCode': 'NAD27      '}))
        self.assertEqual(response.status_code, 401)

    def test_bad_token(self, mock):
        bad_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'bad_secret')
        response = self.app_client.post('/transformer/decimal_location',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(bad_token.decode('utf-8'))},
                                        data=json.dumps({
                                            'latitudee': ' 400000    ',
                                            'longitude': ' 1000000    ',
                                            'coordinateDatumCode': 'NAD27      '}))
        self.assertEqual(response.status_code, 422)


class StationIxTestCase(TestCase):

    def setUp(self):
        app.application.config['JWT_SECRET_KEY'] = 'secret'
        app.application.config['JWT_PUBLIC_KEY'] = None
        app.application.config['JWT_ALGORITHM'] = 'HS256'
        app.application.config['AUTH_TOKEN_KEY_URL'] = ''
        app.application.config['JWT_DECODE_AUDIENCE'] = None
        app.application.testing = True
        self.app_client = app.application.test_client()

    def test_station_name_with_whitespace(self):
        good_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret')
        response = self.app_client.post('/transformer/station_ix',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(good_token.decode('utf-8'))},
                                        data='{"stationName": "Station Name"}')
        self.assertEqual(response.status_code, 200)
        self.assertEqual(json.loads(response.data), {'stationIx': 'STATIONNAME'})

    def test_station_name_with_nonalphanumerics(self):
        good_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret')
        response = self.app_client.post('/transformer/station_ix',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(good_token.decode('utf-8'))},
                                        data='{"stationName": "Station#_Name1$"}')
        self.assertEqual(response.status_code, 200)
        self.assertEqual(json.loads(response.data), {'stationIx': 'STATIONNAME1'})

    def test_invalid_request_payload(self):
        good_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'secret')
        response = self.app_client.post('/transformer/station_ix',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(good_token.decode('utf-8'))},
                                        data='{"stationNam": "Station#_Name1$"}')
        self.assertEqual(response.status_code, 400)

    def test_no_auth_header(self):
        response = self.app_client.post('/transformer/station_ix',
                                        content_type='application/json',
                                        data='{"stationNam": "Station#_Name1$"}')
        self.assertEqual(response.status_code, 401)

    def test_bad_token(self):
        bad_token = jwt.encode({'authorities': ['one_role', 'two_role']}, 'bad_secret')
        response = self.app_client.post('/transformer/station_ix',
                                        content_type='application/json',
                                        headers={'Authorization': 'Bearer {0}'.format(bad_token.decode('utf-8'))},
                                        data='{"stationName": "Station#_Name1$"}')
        self.assertEqual(response.status_code, 422)



utils.py
from pyproj import Proj, transform
from app import application

nad83_proj = Proj('+init=EPSG:4269')
nad27_proj = Proj('+init=EPSG:4267')

DATUM_TO_PROJ = {
    'NAD83' : nad83_proj,
    'NAD27' : nad27_proj,
}

def transform_to_decimal_degrees(degrees, minutes, seconds):
    fdegrees = float(degrees)
    abs_result = abs(fdegrees) + float(minutes) / 60 + float(seconds) / 3600
    return abs_result if fdegrees > 0 else -abs_result

def transform_latitude_to_decimal_degrees(angle):
    result = transform_to_decimal_degrees(angle[0:3], angle[3:5], angle[5:])
    if result > 90.0 or result < -90.0:
        raise ValueError('Latitude angle ' + str(angle) + ' is not in the inclusive range of (-90, 90)')
    return result

def transform_longitude_to_decimal_degrees(angle):
    result = -transform_to_decimal_degrees(angle[0:4], angle[4:6], angle[6:])
    if result > 180.0 or result < -180.0:
        raise ValueError('Longitude angle ' + str(angle) + ' is not in the inclusive range of (-180, 180)')
    return result

def transform_location_to_decimal_location(latitude, longitude, coord_datum):
    datum = coord_datum.rstrip()
    if datum in DATUM_TO_PROJ:
        try:
            dlat = transform_latitude_to_decimal_degrees(latitude)
            dlon = transform_longitude_to_decimal_degrees(longitude)
        except ValueError as err:
            dlat, dlon = None, None
            application.logger.warning("An error occurred while transforming lat/lon to decimal degrees: "
                + str(err))
        else:
            if datum != 'NAD83':
                dlon, dlat = transform(DATUM_TO_PROJ[datum], nad83_proj, dlon, dlat)

    else:
        dlat, dlon = None, None

    return {
        'decimalLatitude' : dlat,
        'decimalLongitude' : dlon
    }


config.py
import os

DEBUG = False

# The following four variables configure authentication

# If using a public key, set the environment variable AUTH_TOKEN_KEY_URL to the url where it can be retrieved
AUTH_TOKEN_KEY_URL = os.getenv('oauth_server_token_key_url')

# Set the path to the certificate for AUTH_TOKEN_KEY_URL. If set to False then, SSL verification will not occur.
AUTH_CERT_PATH = os.getenv('oauth_server_cert_path', True)

# If using the above, use the AUTH_TOKEN_KEY_ALGORITHM to set the algorithm. By default it will be RS256
JWT_ALGORITHM = os.getenv('jwt_algorithm', 'HS256')

# Set the JWT_DECODE_AUDIENCE environment variable to the value of the 'aud' claim in the
JWT_DECODE_AUDIENCE = os.getenv('jwt_decode_audience')

# Configure exception JSON to not always output with a `message` field. Global exception handling is done via a custom handler in services.py.
ERROR_INCLUDE_MESSAGE=False

services.py
import pkg_resources
import re

from flask import request
from flask_restplus import Api, Resource, fields

from werkzeug.exceptions import BadRequest

from app import application
from flask_restplus_jwt import JWTRestplusManager, jwt_required
from utils import transform_location_to_decimal_location

# This will add the Authorize button to the swagger docs
authorizations = {
    'apikey': {
        'type': 'apiKey',
        'in': 'header',
        'name': 'Authorization'
    }
}

api = Api(application,
          title='MLR Legacy Transformer',
          description='Provides services which transform MLR legacy fields',
          default='Transformer',
          doc='/api',
          authorizations=authorizations
          )

# Setup the Flask-JWT-Simple extension
jwt = JWTRestplusManager(api, application)

lat_lon_model = api.model('LatLonModel', {
    'latitude': fields.String,
    'longitude': fields.String,
    'coordinateDatumCode': fields.String,
})
decimal_lat_lon_model = api.model('DecimalLatLonModel', {
    'decimalLatitude': fields.Float,
    'decimalLongitude': fields.Float
})
station_name_model = api.model('StationNameModel', {
    'stationName': fields.String
})
station_ix_model = api.model('StationIxModel', {
    'stationIx': fields.String
})
error_model = api.model('ErrorModel', {
    'error_message': fields.String(required=True)
})

expected_lat_lon_model_keys = set(iter(lat_lon_model.keys()))
expected_station_name_model_keys = set(iter(station_name_model.keys()))


def _handle_missing_keys(json_data, expected_keys):
    """
    :param dict json_data:
    :param set expected_keys:
    :raises BadRequest if any keys in expected_keys are missing from json_data
    """
    request_keys = set(iter(json_data.keys()))
    missing_keys = expected_keys.difference(request_keys)
    if missing_keys:
        raise BadRequest('Missing keys: {0}'.format(', '.join(missing_keys)))


@api.route('/transformer/decimal_location')
class DecimalLocation(Resource):

    @api.response(200, 'Successful', decimal_lat_lon_model)
    @api.response(400, 'Missing keys', error_model)
    @api.response(500, 'Transformeration failed.', error_model)
    @api.response(401, 'Not authorized or token expired', error_model)
    @api.response(422, 'Invalid JWT Token', error_model)
    @api.doc(security='apikey')
    @api.expect(lat_lon_model)
    @jwt_required
    def post(self):
        request_body = request.get_json()
        try:
            _handle_missing_keys(request_body, expected_lat_lon_model_keys)
            decimal_location = transform_location_to_decimal_location(
                request_body.get('latitude'),
                request_body.get('longitude'),
                request_body.get('coordinateDatumCode')
            )
        except BadRequest as err:
            response, status = {'error_message': err.description}, 400
        else:
            response, status = decimal_location, 200

        return response, status 


@api.route('/transformer/station_ix')
class StationIx(Resource):

    @api.response(200, 'Successful', station_ix_model)
    @api.response(400, "Missing keys:", error_model)
    @api.response(500, 'Transformeration failed.', error_model)
    @api.response(401, 'Not authorized or token expired', error_model)
    @api.response(422, 'Invalid JWT Token', error_model)
    @api.doc(security='apikey')
    @api.expect(station_name_model)
    @jwt_required
    def post(self):
        request_body = request.get_json()
        try:
            _handle_missing_keys(request_body, expected_station_name_model_keys)
            station_ix = re.sub('\s|[^a-zA-Z0-9]', '', request_body.get('stationName'))
        except BadRequest as err:
            response, status = {'error_message': err.description}, 400
        else:
            response,status = {'stationIx': station_ix.upper()}, 200

        return response, status


version_model = api.model('VersionModel', {
    'version': fields.String,
    'artifact': fields.String
})


@api.route('/version')
class Version(Resource):

    @api.response(200, 'Success', version_model)
    def get(self):
        try:
            distribution = pkg_resources.get_distribution('usgs_wma_mlr_legacy_transformer')
        except pkg_resources.DistributionNotFound:
            resp = {
                "version": "local_development",
                "artifact": None
            }
        else:
            resp = {
                "version": distribution.version,
                "artifact": distribution.project_name
            }
        return resp

@api.errorhandler
def default_error_handler(error):
    '''Default error handler'''
    return {'error_message': str(error)}, getattr(error, 'code', 500)